scripts/template_matcher.py

```diff
@@ -19,55 +20,90 @@ def __init__ (self, images, min_match_count=10, good_thresh=0.7):
         returns a dictionary, keys being signs, values being confidences
         """
         visual_diff = {}
-
-        # TODO: get keypoints and descriptors from input image using SIFT
+        template_confidence = {}
+        # TODONE: get keypoints and descriptors from input image using SIFT
         #       store keypoints in variable kp and descriptors in des
-
+        kp, des = self.sift.detectAndCompute(img,None)
+        
         for k in self.signs.keys():
             #cycle trough templage images (k) and get the image differences
             visual_diff[k] = self._compute_prediction(k, img, kp, des)
 
         if visual_diff:
-            pass
-            # TODO: convert difference between images (from visual_diff)
+            # TODONE: convert difference between images (from visual_diff)
             #       to confidence values (stored in template_confidence)
+            for k in visual_diff:
+                template_confidence[k] = 1/visual_diff[k]
 
+            factor = 1.0/sum(template_confidence.itervalues())
+            for k in template_confidence:
+                template_confidence[k] *= factor
+        
         else: # if visual diff was not computed (bad crop, homography could not be computed)
             # set 0 confidence for all signs
             template_confidence = {k: 0 for k in self.signs.keys()}
             
-        #TODO: delete line below once the if statement is written
-        template_confidence = {k: 0 for k in self.signs.keys()}
-
         return template_confidence
 
-
     def _compute_prediction(self, k, img, kp, des):
         """
         Return comparison values between a template k and given image
         k: template image for comparison, img: scene image
         kp: keypoints from scene image,   des: descriptors from scene image
         """
 
-        # TODO: find corresponding points in the input image and the templae image
+        # TODONE: find corresponding points in the input image and the template image
         #       put keypoints from template image in template_pts
         #       put corresponding keypoints from input image in img_pts
-
-
-        #TODO: change img to img_T once you do the homography transform
-        visual_diff = compare_images(img, self.signs[k])
+        
+        # BFMatcher with default params
+        matches = self.bf.knnMatch(self.descs[k], des, k=2)
+        # Apply ratio test and add the matches with high probability (low matches)
+        good = []
+        for m,n in matches:
+            if m.distance < 0.75*n.distance:
+                good.append(m)
+
+        # Code from: http://opencv-python-tutroals.readthedocs.io/en/latest/py_tutorials/py_feature2d/py_feature_homography/py_feature_homography.html
+        img_pts = np.float32([ kp[m.trainIdx].pt for m in good ]).reshape(-1,1,2)
+        template_pts = np.float32([ self.kps[k][m.queryIdx].pt for m in good ]).reshape(-1,1,2)
```

Comment:
This all looks good, and probably works, but you should make sure you understand what this code does